File: go-algorithm/leetcode/problem211.go
package leetcode

// 211. Design Add and Search Words Data Structure
// https://leetcode.com/problems/design-add-and-search-words-data-structure/

type WordDictionary struct {
	root *TrieNode
}

func _() WordDictionary {
	return WordDictionary{&TrieNode{}}
}

func (d *WordDictionary) AddWord(word string) {
	d.root.Insert(word)
}

func (d *WordDictionary) Search(word string) bool {
	var dfs func(int, *TrieNode) bool
	dfs = func(i int, node *TrieNode) bool {
		if i == len(word) {
			return node.word == word
		}
		ch := word[i]
		if ch != '.' {
			child := node.children[ch-'a']
			if child != nil && dfs(i+1, child) {
				return true
			}
		} else {
			for _, child := range node.children {
				if child != nil && dfs(i+1, child) {
					return true
				}
			}
		}
		return false
	}
	return dfs(0, d.root)
}

File: java-algorithm/src/main/java/com/leetcode/Problem211.java
package com.leetcode;

/**
 * 211. Design Add and Search Words Data Structure
 * https://leetcode.com/problems/design-add-and-search-words-data-structure/
 */
public class Problem211 {
    static class WordDictionary {
        private Trie t;

        public WordDictionary() {
            this.t = new Trie();
        }

        public void addWord(String word) {
            Trie cur = t;
            for (int i = 0; i < word.length(); i++) {
                int c = word.charAt(i) - 'a';
                if (cur.children[c] == null) {
                    cur.children[c] = new Trie();
                }
                cur = cur.children[c];
            }
            cur.end = true;
        }

        public boolean search(String word) {
            return t.search(word);
        }
    }

    static class Trie {
        private Trie[] children = new Trie[26];
        private boolean end = false;

        public Trie() {
        }

        public boolean search(String word) {
            Trie t = this;
            for (int i = 0; i < word.length(); i++) {
                char c = word.charAt(i);
                if (c == '.') {
                    for (Trie next : t.children) {
                        if (next != null && next.search(word.substring(i + 1))) {
                            return true;
                        }
                    }
                    return false;
                }
                if (t.children[c - 'a'] == null) {
                    return false;
                }
                t = t.children[c - 'a'];
            }
            return t.end;
        }
    }

    public static void main(String[] args) {
        Problem211.WordDictionary wordDictionary = new Problem211.WordDictionary();
        wordDictionary.addWord("bad");
        wordDictionary.addWord("dad");
        wordDictionary.addWord("mad");
        System.out.println(wordDictionary.search("b.."));
    }
}
